AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,886 @@
+//
+//  ExampleAVAudioEngineDevice.m
+//  AudioDeviceExample
+//
+//  Copyright © 2018 Twilio, Inc. All rights reserved.
+//
+
+#import "ExampleAVAudioEngineDevice.h"
+
+// We want to get as close to 10 msec buffers as possible because this is what the media engine prefers.
+static double const kPreferredIOBufferDuration = 0.01;
+
+// We will use mono playback and recording where available.
+static size_t const kPreferredNumberOfChannels = 1;
+
+// An audio sample is a signed 16-bit integer.
+static size_t const kAudioSampleSize = 2;
+static uint32_t const kPreferredSampleRate = 48000;
+
+/*
+ * Calls to AudioUnitInitialize() can fail if called back-to-back after format change or add remove tracks
+ * audio deivce instances. A fall-back solution is to allow multiple sequential calls
+ * with as small delay between each. This factor sets the max number of allowed
+ * initialization attempts.
+ */
+static const int kMaxNumberOfAudioUnitInitializeAttempts = 5;
+
+// Audio renderer contexts used in core audio's playout callback to retrieve the sdk's audio device context.
+typedef struct AudioRendererContext {
+    // Audio device context received in AudioDevice's `startRendering:context` callback.
+    TVIAudioDeviceContext deviceContext;
+
+    // Maximum frames per buffer.
+    size_t maxFramesPerBuffer;
+
+    // Buffer passed to AVAudioEngine's manualRenderingBlock to receive the mixed audio data.
+    AudioBufferList *bufferList;
+
+    /*
+     * Points to AVAudioEngine's manualRenderingBlock. This block is called by VoiceProcessingIO playout call back in
+     * to receive the mixed audio data from AVAudioEngine in real time.
+     */
+    void *renderBlock;
+} AudioRendererContext;
```

Comment:
If its possible to declare that this class is only available in iOS 11 then I would prefer to use the strong type for renderBlock.